pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,450 @@
+	if logger == nil {
+		logger = logging.ConsoleErrorLogger
+	}
+
+	return &Transport{
+		ownID:            ownID,
+		host:             h,
+		incomingMessages: make(chan *events.EventList),
+		logger:           logger,
+		conns:            make(map[types.NodeID]*connInfo),
+		stopChan:         make(chan struct{}),
+		streamChan:       make(chan *newStream, 1),
+	}, nil
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, "starting libp2p transport", "src", t.ownID, "listen", t.host.Addrs())
+	t.runStreamUpdater()
+	t.host.SetStreamHandler(ProtocolID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "stopping libp2p transport", "src", t.ownID)
+	defer t.logger.Log(logging.LevelDebug, "libp2p transport stopped ", "src", t.ownID)
+
+	close(t.stopChan)
+
+	t.host.RemoveStreamHandler(ProtocolID)
+
+	t.connsLock.Lock()
+	for nodeID, c := range t.conns {
+		if err := c.Stream.Close(); err != nil {
+			t.logger.Log(logging.LevelError, "could not close stream", "src", t.ownID, "dst", nodeID, "err", err)
+		}
+		delete(t.conns, nodeID)
+	}
+	t.connsLock.Unlock()
+}
+
+func (t *Transport) CloseOldConnections(newNodes map[types.NodeID]types.NodeAddress) {
+	t.connsLock.Lock()
+	defer t.connsLock.Unlock()
+
+	for nodeID, c := range t.conns {
+		// Close the old stream to a node.
+		if _, foundInNewNodes := newNodes[nodeID]; !foundInNewNodes {
+			t.logger.Log(logging.LevelDebug, "closing old connection", "src", t.ownID, "dst", nodeID)
+
+			if err := c.Stream.Close(); err != nil {
+				t.logger.Log(logging.LevelError, "could not close old stream to node", "src", t.ownID, "dst", nodeID, "err", err)
+			}
+
+			delete(t.conns, nodeID)
+		}
+	}
+}
+
+func (t *Transport) Connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	if len(nodes) == 0 {
+		t.logger.Log(logging.LevelWarn, "no nodes to connect to")
+		return
+	}
+
+	parsedAddrInfo := make(map[types.NodeID]*peer.AddrInfo)
+	for nodeID, addr := range nodes {
+		info, err := peer.AddrInfoFromP2pAddr(addr)
+		if err != nil {
+			t.logger.Log(logging.LevelWarn, "connect: failed to parse addr", "src", t.ownID, "dest", nodeID, "addr", addr, "err", err)
+			continue
+		}
+		parsedAddrInfo[nodeID] = info
+	}
+
+	t.connsLock.Lock()
+	for nodeID := range parsedAddrInfo {
+		if nodeID == t.ownID {
+			continue
+		}
+		_, found := t.conns[nodeID]
+		if !found {
+			t.conns[nodeID] = &connInfo{AddrInfo: parsedAddrInfo[nodeID], Stream: nil}
+		}
+	}
+	t.connsLock.Unlock()
+
+	t.connect(ctx, maputil.GetKeys(parsedAddrInfo))
+}
+
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, msg *messagepb.Message) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-t.stopChan:
+		return fmt.Errorf("transport stopped")
+	default:
+	}
+
+	out, err := t.encode(msg)
+	if err != nil {
+		return err
+	}
+
+	err = t.sendPayload(dest, out)
+	if err != nil {
+		t.connect(ctx, []types.NodeID{dest})
```

Comment:
I'm a bit worried that, if the connection breaks and the new one takes some time to establish, there might be literally thousands of messages that fail to be sent in the meantime, each of them triggering a new connection attempt. All of these attempts (but one) will be closed almost immediately. Could that happen?

Reply:
In fact, nevermind, let's do this in a separate PR.